cranelift/interpreter/src/interpreter.rs

```diff
@@ -720,4 +929,51 @@ mod tests {
 
         assert_eq!(trap, CraneliftTrap::User(TrapCode::HeapOutOfBounds));
     }
+
+    /// Most heap tests are in .clif files using the filetest machinery. However, this is a sanity
+    /// check that the heap mechanism works without the rest of the filetest infrastructure
+    #[test]
+    fn heap_sanity_test() {
+        let code = "
+        function %heap_load_store(i64 vmctx) -> b1 {
+            gv0 = vmctx
+            gv1 = load.i64 notrap aligned gv0+0
+            ; gv2/3 do nothing, but makes sure we understand the iadd_imm mechanism
+            gv2 = iadd_imm.i64 gv1, 1
+            gv3 = iadd_imm.i64 gv2, -1
+            heap0 = static gv3, min 0x1000, bound 0x1_0000_0000, offset_guard 0, index_type i64
+
+        block0(v0: i64):
+            v1 = iconst.i64 0
+            v2 = iconst.i64 123
+            v3 = heap_addr.i64 heap0, v1, 8
+            store.i64 v2, v3
+            v4 = load.i64 v3
+            v5 = icmp eq v2, v4
+            return v5
+        }";
+
+        let func = parse_functions(code).unwrap().into_iter().next().unwrap();
+        let mut env = FunctionStore::default();
+        env.add(func.name.to_string(), &func);
+        let mut state = InterpreterState::default().with_function_store(env);
+
+        let heap0 = state.register_heap(HeapInit::Zeroed(0x1000));
+        let base_addr = state.get_heap_address(I64, heap0, 0).unwrap();
+
+        // Build a vmctx struct by writing the base pointer at index 0
+        let mut vmctx_struct = vec![0u8; 8];
+        base_addr.write_to_slice(&mut vmctx_struct[..]);
+
+        // This is our vmctx "heap"
+        let vmctx = state.register_heap(HeapInit::FromBacking(vmctx_struct));
+        let vmctx_addr = state.get_heap_address(I64, vmctx, 0).unwrap();
```

Comment:
I'm not too happy with this interface as a way to register a VMContext struct.

We need to have a way to address this struct, and currently we only have Stack and Heap addresses. Adding a whole new region just for VMContext structs seems a bit much, but this is also not very clear.